src/Select/SelectHelper.php

```diff
@@ -0,0 +1,141 @@
+
+            if ($fieldDefinition) {
+                foreach (self::DIRECTIVES as $directiveType) {
+                    if ($directive = ASTHelper::directiveDefinition($fieldDefinition, $directiveType)) {
+                        assert($directive instanceof DirectiveNode);
+
+                        if (in_array($directiveType, self::DIRECTIVES_RETURN)) {
+                            return [];
+                        }
+
+                        if (in_array($directiveType, self::DIRECTIVES_REQUIRING_LOCAL_KEY)) {
+                            $relationName = ASTHelper::directiveArgValue($directive, 'relation', $field);
+
+                            if (method_exists($model, $relationName)) {
+                                $relation = $model->{$relationName}();
+
+                                $localKey = AppVersion::below(5.7)
+                                    ? Utils::accessProtected($relation, 'localKey')
+                                    : $relation->getLocalKeyName();
+
+                                $selectColumns[] = $localKey;
+                            }
+                        }
+
+                        if (in_array($directiveType, self::DIRECTIVES_REQUIRING_FOREIGN_KEY)) {
+                            $relationName = ASTHelper::directiveArgValue($directive, 'relation', $field);
+
+                            if (method_exists($model, $relationName)) {
+                                $foreignKey = AppVersion::below(5.8)
+                                    ? $model->{$relationName}()->getForeignKey()
+                                    : $model->{$relationName}()->getForeignKeyName();
+
+                                $selectColumns[] = $foreignKey;
+                            }
+                        }
+
+                        continue 2;
+                    }
+                }
+
+                if ($directive = ASTHelper::directiveDefinition($fieldDefinition, 'select')) {
+                    // append selected columns in select directive to selection
+                    $selectFields = ASTHelper::directiveArgValue($directive, 'columns', []);
+                    $selectColumns = array_merge($selectColumns, $selectFields);
+                } elseif ($directive = ASTHelper::directiveDefinition($fieldDefinition, 'rename')) {
+                    // append renamed attribute to selection
+                    $renamedAttribute = ASTHelper::directiveArgValue($directive, 'attribute');
+                    $selectColumns[] = $renamedAttribute;
+                } else {
+                    // fallback to selecting the field name
+                    $selectColumns[] = $field;
+                }
+            }
+        }
+
+        /** @var array<int, string> $selectColumns */
+        $selectColumns = array_filter($selectColumns, function ($column) use ($model): bool {
+            return ! $model->hasGetMutator($column) && ! method_exists($model, $column);
```

Comment:
I don't think this condition is sufficient to determine if a field is a column.

```graphql
type Foo {
  foo: String @method(name: "bar") # falsely assumed to be a column due to not recognizing the directive
  bar: String # falsely assumed not to be a column due to the presence of method bar()
  baz: String # falsely assumed not to be a column due to the presence of a get mutator
}
```

```php
final class Foo extends Model {
    public function bar(): string { ... }
    public function getBazAttribute(): string {
        return $this->attributes['baz'];
    }
}
```

Given the existence of custom field resolver directives, I don't see how we could reliably determine if a field is a column without explicit configuration.

Reply:
Good example. I removed it and only checked if the method existed for the relation. And by default, select the primary key to avoid no column selected if the method exists and can't find any select column. Then, we can use `@select` if any other columns need to be chosen.